include/deal.II/lac/vector.templates.h

```diff
@@ -614,6 +563,487 @@ namespace internal
+  {
+    const size_type vec_size = last - first;
+    if (vec_size <= 4096)
+      {
+        // the vector is short enough so we perform the summation. first
+        // work on the regular part. The innermost 32 values are expanded in
+        // order to obtain known loop bounds for most of the work.
+        size_type index = first;
+        ResultType outer_results [128];
+        size_type n_chunks = vec_size / 32;
+        const size_type remainder = vec_size % 32;
+        Assert (remainder == 0 || n_chunks < 128, ExcInternalError());
+
+        // Select between the regular version and vectorized version based
+        // on the number types we are given. To choose the vectorized
+        // version often enough, we need to have all tasks but the last one
+        // to be divisible by the vectorization length
+        accumulate_regular(op, n_chunks, index, outer_results,
+                           internal::bool2type<Operation::vectorizes>());
+
+        // now work on the remainder, i.e., the last up to 32 values. Use
+        // switch statement with fall-through to work on these values.
+        if (remainder > 0)
+          {
+            AssertIndexRange(n_chunks, 129);
+            const size_type inner_chunks = remainder / 8;
+            Assert (inner_chunks <= 3, ExcInternalError());
+            const size_type remainder_inner = remainder % 8;
+            ResultType r0 = ResultType(), r1 = ResultType(),
+                       r2 = ResultType();
+            switch (inner_chunks)
+              {
+              case 3:
+                r2 = op(index++);
+                for (size_type j=1; j<8; ++j)
+                  r2 += op(index++);
+              // no break
+              case 2:
+                r1 = op(index++);
+                for (size_type j=1; j<8; ++j)
+                  r1 += op(index++);
+                r1 += r2;
+              // no break
+              case 1:
+                r2 = op(index++);
+                for (size_type j=1; j<8; ++j)
+                  r2 += op(index++);
+              // no break
+              default:
+                for (size_type j=0; j<remainder_inner; ++j)
+                  r0 += op(index++);
+                r0 += r2;
+                r0 += r1;
+                if (n_chunks == 128)
+                  outer_results[127] += r0;
+                else
+                  {
+                    outer_results[n_chunks] = r0;
+                    n_chunks++;
+                  }
+                break;
+              }
+          }
+        AssertDimension(index, last);
+
+        // now sum the results from the chunks
+        // recursively
+        while (n_chunks > 1)
+          {
+            if (n_chunks % 2 == 1)
+              outer_results[n_chunks++] = ResultType();
+            for (size_type i=0; i<n_chunks; i+=2)
+              outer_results[i/2] = outer_results[i] + outer_results[i+1];
+            n_chunks /= 2;
```

Comment:
No, this is direct copy from the current code. Whether it's necessary: Pairwise summation gives in the statistical sense the smallest error of the various summation techniques, therefore I selected it back in 2012. Certainly at this point you probably only lose one or two bits at most (the length is up to 128) by linear summation and we still have plenty of pairwise operations further up the chain, but it's not that terribly complicated either in this way.